Changes in module 'nf-core/multiqc'
--- modules/nf-core/multiqc/main.nf
+++ modules/nf-core/multiqc/main.nf
@@ -1,10 +1,10 @@
 process MULTIQC {
-    label 'process_single'
+    label 'process_medium'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/multiqc:1.24.1--pyhdfd78af_0' :
-        'biocontainers/multiqc:1.24.1--pyhdfd78af_0' }"
+        'docker.io/multiqc/multiqc:v1.24.1' }"
 
     input:
     path  multiqc_files, stageAs: "?/*"

************************************************************
