Changes in module 'nf-core/spaceranger/count'
--- modules/nf-core/spaceranger/count/meta.yml
+++ modules/nf-core/spaceranger/count/meta.yml
@@ -83,7 +83,7 @@
   - outs:
       type: file
       description: Files containing the outputs of Cell Ranger, see official 10X Genomics documentation for a complete list
-      pattern: "${meta.id}/outs/*"
+      pattern: "outs/*"
   - versions:
       type: file
       description: File containing software versions

--- modules/nf-core/spaceranger/count/main.nf
+++ modules/nf-core/spaceranger/count/main.nf
@@ -10,7 +10,7 @@
     path(probeset)
 
     output:
-    tuple val(meta), path("**/outs/**"), emit: outs
+    tuple val(meta), path("outs/**"), emit: outs
     path "versions.yml", emit: versions
 
     when:
@@ -46,6 +46,7 @@
         $alignment \\
         $slidefile \\
         $args
+    mv ${prefix}/outs outs
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
@@ -58,10 +59,9 @@
     if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
         error "SPACERANGER_COUNT module does not support Conda. Please use Docker / Singularity / Podman instead."
     }
-    def prefix = task.ext.prefix ?: "${meta.id}"
     """
-    mkdir -p "${prefix}/outs/"
-    touch ${prefix}/outs/fake_file.txt
+    mkdir -p outs/
+    touch outs/fake_file.txt
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
